File: validNumber.cpp
class Solution {
public:
    bool isNumber(string s) {
        int p = 0;
        skipSpace(s, &p);
        skipSign(s, &p);
        if(!hasNumber(s, &p)) return false;
        
        if(s[p]!= '\0' && (s[p] == 'e' || s[p] == 'E')){
            p++;
            skipSign(s, &p);
            if(!hasInteger(s, &p)) return false;
        }
        skipSpace(s, &p);
        return (s[p] == '\0');
    }
private:
    void skipSpace(string s, int *p){
        while(s[*p]!='\0' && s[*p] == ' ') (*p)++;
    }
    void skipSign(string s, int *p){
        while(s[*p] != '\0' && (s[*p] == '+' || s[*p] == '-')) (*p)++;
    }
    bool hasInteger(string s, int *p){
        int len = 0;
        while(s[*p] != '\0' && isdigit(s[*p])) {(*p)++; len++;}
        return len!=0;
    }
    bool hasNumber(string s, int *p){
        int len = 0;
        while(s[*p] != '\0' && isdigit(s[*p])) {(*p)++; len++;}
        if (s[*p] != '\0' && s[*p]=='.') (*p)++;
        while(s[*p] != '\0' && isdigit(s[*p])) {(*p)++; len++;}
        return len!=0;
    }
    
};

File: validNumber.py
class Solution:
    # @param s {string}
    # @return {boolean}
    def isNumber(self, s):
        state = 0
        for i in xrange(0,len(s)):
            state = self.nextState(state, s[i])
            if state == -1: 
                return False
        state = self.nextState(state, ' ')
        return state == 8
    
    def nextState(self, state, char):
        transititionTable = [[0,2,1,3,-1,-1], [-1,2,-1,3,-1,-1], [8,2,-1,4,5,-1], [-1,4,-1,-1,-1,-1], [8,4,-1,-1,5,-1],[-1,7,6,-1,-1,-1],[-1,7,-1,-1,-1,-1], [8,7,-1,-1,-1,-1], [8,-1,-1,-1,-1,-1]];
        return transititionTable[state][self.getSymbol(char)]
    
    def getSymbol(self, char):
        if char == ' ' or char == '\t':
            return 0
        if char.isdigit():
            return 1
        if char == '+'  or char == '-':
            return 2
        if char == '.':
            return 3
        if char == 'E' or char =='e':
            return 4
        return 5